arviz/kde_utils.py

```diff
@@ -0,0 +1,895 @@
+
+
+def kde_linear(
+    x,
+    bw="experimental",
+    adaptive=False,
+    extend=False,
+    bound_correction=True,
+    extend_fct=0,
+    bw_fct=1,
+    bw_return=False,
+    custom_lims=None,
+    cumulative=False,
+    grid_len=512
+):
+    """
+    1 dimensional density estimation for linear data.
+
+    Given an array of data points `x` it returns an estimate of
+    the probability density function that generated the samples in `x`.
+
+    Parameters
+    ----------
+    x : 1D numpy array
+        Data used to calculate the density estimation.
+        Theoritically it is a random sample obtained from $f$,
+        the true probability density function we aim to estimate.
+    bw: int, float or str, optional
+        If numeric, indicates the bandwidth and must be positive.
+        If str, indicates the method to estimate the bandwidth and must be
+        one of "scott", "silverman", "isj" or "experimental".
+        Defaults to "experimental".
+    adaptive: boolean, optional
+        Indicates if the bandwidth is adaptative or not.
+        It is the recommended approach when there are multiple modalities
+        with different spread.
+        It is not compatible with convolution. Defaults to False.
+    extend: boolean, optional
+        Whether to extend the observed range for `x` in the estimation.
+        It extends each bound by a multiple of the standard deviation of `x`
+        given by `extend_fct`. Defaults to False.
+    bound_correction: boolean, optional
+        Whether to perform boundary correction on the bounds of `x` or not.
+        Defaults to True.
+    extend_fct: float, optional
+        Number of standard deviations used to widen the
+        lower and upper bounds of `x`. Defaults to 0.5.
+    bw_fct: float, optional
+        A value that multiplies `bw` which enables tuning smoothness by hand.
+        Must be positive. Values below 1 decrease smoothness while values
+        above 1 decrease it. Defaults to 1 (no modification).
+    bw_return: bool, optional
+        Whether to return the estimated bandwidth in addition to the
+        other objects. Defaults to False.
+    custom_lims: list or tuple, optional
+        A list or tuple of length 2 indicating custom bounds
+        for the range of `x`. Defaults to None which disables custom bounds.
+    cumulative: bool, optional
+        Whether return the PDF or the cumulative PDF. Defaults to False.
+    grid_len: int, optional
+        The number of intervals used to bin the data points
+        (a.k.a. the length of the grid used in the estimation)
+        Defaults to 512.
+
+    Returns
+    -------
+    grid : Gridded numpy array for the x values.
+    pdf : Numpy array for the density estimates.
+    bw: optional, the estimated bandwidth.
+    """
+
+    # Check `x` is from appropiate type
+    x = check_type(x)
+
+    # Assert `bw_fct` is numeric and positive
+    assert isinstance(bw_fct, (int, float))
+    assert bw_fct > 0
```

Comment:
similar comment to one above, these two should probably be type error and value error respectively, right now they'd raise an assertion error without any custom message

Reply:
the isinstance int, float will also give problems with numpy dtypes i.e. float32

Reply:
I can replace it with the following

```python
if not isinstance(bw_fct, (int, float, np.integer, np.floating)):
    raise TypeError(f"`bw_fct` must be a positive number, not an object of {type(bw_fct)}.")
    
if not bw_fct > 0:
    raise ValueError(f"`bw_fct` must be a positive number, not {bw_fct}.")
```

Reply:
Doesn't floating work with python ints too?

Reply:
Is this what you mean? It returns `False`
```python
isinstance(1, np.floating)
# False
```

Reply:
Haha, my bad: with python floats